File: src/main/java/org/jukeboxmc/network/handler/MobEquipmentHandler.java
package org.jukeboxmc.network.handler;

import org.cloudburstmc.protocol.bedrock.packet.MobEquipmentPacket;
import org.jukeboxmc.Server;
import org.jukeboxmc.inventory.Inventory;
import org.jukeboxmc.inventory.PlayerInventory;
import org.jukeboxmc.inventory.WindowId;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.player.Player;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class MobEquipmentHandler implements PacketHandler<MobEquipmentPacket> {

    @Override
    public void handle( MobEquipmentPacket packet, Server server, Player player ) {
        Inventory inventory = this.getInventory( player, WindowId.getWindowIdById( packet.getContainerId() ) );
        if ( inventory != null ) {
            Item item = inventory.getItem( packet.getHotbarSlot() );
            if ( !item.equals( new Item( packet.getItem(), false) ) ) {
                inventory.sendContents( player );
                return;
            }
            if ( inventory instanceof PlayerInventory playerInventory ) {
                playerInventory.setItemInHandSlot( packet.getHotbarSlot() );
                player.setAction( false );
            }
        }
    }

    private Inventory getInventory( Player player, WindowId windowId ) {
        return switch ( windowId ) {
            case PLAYER -> player.getInventory();
            case CURSOR_DEPRECATED, CURSOR -> player.getCursorInventory();
            case ARMOR_DEPRECATED, ARMOR -> player.getArmorInventory();
            default -> player.getCurrentInventory();
        };
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/network/handler/MobEquipmentHandler.kt
package org.jukeboxmc.server.network.handler

import org.cloudburstmc.protocol.bedrock.packet.MobEquipmentPacket
import org.jukeboxmc.api.inventory.Inventory
import org.jukeboxmc.server.JukeboxServer
import org.jukeboxmc.server.extensions.toJukeboxItem
import org.jukeboxmc.server.inventory.JukeboxInventory
import org.jukeboxmc.server.inventory.JukeboxPlayerInventory
import org.jukeboxmc.server.inventory.WindowId
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer

class MobEquipmentHandler : PacketHandler<MobEquipmentPacket> {

    override fun handle(packet: MobEquipmentPacket, server: JukeboxServer, player: JukeboxPlayer) {
        val windowId = WindowId.getWindowId(packet.containerId)?: return
        val inventory = getInventory(player, windowId)
        if (inventory != null) {
            val item = inventory.getItem(packet.hotbarSlot).toJukeboxItem()
            if (item != JukeboxItem(packet.item, false)) {
                (inventory as JukeboxInventory).sendContents(player)
                return
            }
            if (inventory is JukeboxPlayerInventory) {
                inventory.setItemInHandSlot(packet.hotbarSlot)
                player.setAction(false)
            }
        }
    }

    private fun getInventory(player: JukeboxPlayer, windowId: WindowId): Inventory? {
        return when (windowId) {
            WindowId.PLAYER -> player.getInventory()
            WindowId.CURSOR_DEPRECATED, WindowId.CURSOR -> player.getCursorInventory()
            WindowId.ARMOR_DEPRECATED, WindowId.ARMOR -> player.getArmorInventory()
            else -> player.getCurrentInventory()
        }
    }
}